### PR TITLE
Build cluster setup demystified

### PR DESCRIPTION
**Pull request by listx**
Putting this up for initial review before I add too much (incorrect) technical details. I will probably also add a (simplified?) diagram WRT `gencred` and `kubernetes-external-secrets`, but that can come later.

/cc @chaodaiG

### CODE REVIEW

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:2]: Note that for prow, a build cluster could be any k8s cluster, could be GKE, AKS, or EKS. I think it's fine to only document GKE build cluster for now. But should also keep in mind that build cluster from other providers might also be documented by someone else.

**Review by chaodaiG — COMMENTED** [review]
Overall LGTM, left two comments

File: site/content/en/docs/build-clusters.md
Comment (listx) [inline site/content/en/docs/build-clusters.md:59]: @chaodaiG Do you know how long this is? was it 2 days?

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md:2]: Exactly.

**Comment by listx** [comment]
@chaodaiG @cjwagner Can you have a double-take on for the diagram preview here? https://deploy-preview-17--k8s-prow.netlify.app/docs/build-clusters/

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:64]: 2 days max

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:64]: To be more accurate, 2 days for GKE, the value might be different for other k8s providers

**Comment by chaodaiG** [comment]
> @chaodaiG @cjwagner Can you have a double-take on for the diagram preview here? https://deploy-preview-17--k8s-prow.netlify.app/docs/build-clusters/#grant-gcs-access-for-your-jobs-with-workload-identity

might just be me, although knowing the contexts already, it still feels not very easy to understand the diagram. what do you think about describing the sequence of events with bullets? something like:

1. KSA associated short-lived token generated by gencred
2. token mount onto Prow components, such as PCM
3. PCM runs `kubectl apply` equivalent that is authenticated with the token

**Comment by listx** [comment]
> Member

I can add those things, although I don't think adding numbers to indicate the sequence is a good idea (because not all of the arrows should be numbered).

**Comment by listx** [comment]
@chaodaiG QQ: why do we need this line https://github.com/kubernetes/test-infra/blob/a25542c55d8452bc470f847421c53a6b1710f84c/prow/create-build-cluster.sh#L270 ? Is it because the `gencred` binary needs it when it runs inside the `ci-test-infra-gencred-refresh-kubeconfig` job inside the trusted cluster? If so, does gencred spin up a new container inside the user-provided build cluster to generate the 2-day token and store it in the user's GCP Secret Manager?

**Comment by chaodaiG** [comment]
> @chaodaiG QQ: why do we need this line https://github.com/kubernetes/test-infra/blob/a25542c55d8452bc470f847421c53a6b1710f84c/prow/create-build-cluster.sh#L270 ? Is it because the `gencred` binary needs it when it runs inside the `ci-test-infra-gencred-refresh-kubeconfig` job inside the trusted cluster?

yes

> If so, does gencred spin up a new container inside the user-provided build cluster to generate the 2-day token and store it in the user's GCP Secret Manager?

It does not. `gencred` interacts directly with the api server of build clusters to generate the token

**Comment by listx** [comment]
Can you go into a little more detail about why the gencred-refresher GSA needs `--role="roles/container.admin"` on the build cluster's GCP project? For example, what container does it need to create/delete for it to run?

**Comment by chaodaiG** [comment]
> Can you go into a little more detail about why the gencred-refresher GSA needs `--role="roles/container.admin"` on the build cluster's GCP project? For example, what container does it need to create/delete for it to run?

I see the confusion, this is one of the predefined GKE roles for managing GKE https://cloud.google.com/kubernetes-engine/docs/how-to/iam#predefined, they are recommended roles for interacting with GKE clusters

**Comment by listx** [comment]
> I see the confusion, this is one of the predefined GKE roles for managing GKE https://cloud.google.com/kubernetes-engine/docs/how-to/iam#predefined, they are recommended roles for interacting with GKE clusters

IIUC, `gencred` binary needs to make the build cluster's K8s API server generate a new token for the "root" (cluster-admin KSA role) KSA on the build cluster. `gencred` does this by running as a GSA that has the `roles/container.admin` role defined by GCP. Is this correct?

**Comment by chaodaiG** [comment]
> > I see the confusion, this is one of the predefined GKE roles for managing GKE https://cloud.google.com/kubernetes-engine/docs/how-to/iam#predefined, they are recommended roles for interacting with GKE clusters
> 
> IIUC, `gencred` binary needs to make the build cluster's K8s API server generate a new token for the "root" (cluster-admin KSA role) KSA on the build cluster. `gencred` does this by running as a GSA that has the `roles/container.admin` role defined by GCP. Is this correct?

correct https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L70

**Comment by listx** [comment]
PTAL

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:16]: This is not necessarily true. For example "default" cluster in k8s prow is the build cluster located in `k8s-prow-builds` project instead of `k8s-prow` project. Consider something like:

> By default Prow will schedule jobs in the cluster that maps to kubeconfig alias "default" (Imagine running `kubectl config set-context "default" --cluster=<CLUSTER_CONTEXT>`, the <CLUSTER_CONTEXT> could be any cluster)

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:16]: Speaking of this, I think this document should also include infos related to context aliases. For example a GKE cluster named `cluster-foo` in project `project-bar` under `us-central1` could be named something totally unrelated, such as `the-most-awesome-build-cluster`. And whenever a user wants to configure something for this cluster they should use `the-most-awesome-build-cluster` instead of `cluster-foo`

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:87]: the GCS bucket could be configured on several different levels in Prow, and is not the same for a single Prow instance, for example a user can configure bucket for a build cluster (https://github.com/GoogleCloudPlatform/oss-test-infra/blob/d46723826c93a03f1d6e30d603bd5d0b0f723212/prow/oss/config.yaml#L138). Rarely user might want to use a different bucket for a job, which is also allowed

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:58]: KSA is not a hard requirement, it's just an easier way to generate kubeconfig for authenticating with a cluster. Being able to manage `Pod` resource on the build cluster might be more accurate, for more detail see gencred code

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:61]: not `jobs`, `prowjob` CR doesn't exist in build clusters

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [inline site/content/en/docs/build-clusters.md:76]: it's in the GCP project where prow is in, so the GSA doesn't need GCP secret manager ACL in build cluster project

**Review by chaodaiG — CHANGES_REQUESTED** [review]

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md:16]: I don't want to make this doc any longer. Let me try rephrasing and/or deleting the sentence you mentioned.

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:16]: if you meant to not extending this PR I would agree. Otherwise imo the `alias` vs real `cluster_name` is where our users got tripped most, and it's worth documented

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md:58]: But what about https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78-L79 ? Seems to contradict what you just said. Unless there's other `gencred` codepaths involved here.

I will simply include your words verbatim into the doc.

Comment (listx) [reply site/content/en/docs/build-clusters.md:58]: I'm confused because this line seems to condtradict what you're saying: https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78

But also, I'll include your comment in the doc directly, because it will be faster than me reading through gencred.

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:58]: I don't think there is a conflict here. The point I'd like to make, is that "it must be able to authenticate as a KSA" is not an accurate statement, what Prow really must have is "it must be able to `kubectl get pods`, `kubectl create pods`, `kubectl * pods` etc", and other than creating a KSA, there are multiple other ways this can be achieved. At least gencred supports two different ways https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/cmd/gencred/main.go#L236

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:58]: > I'm confused because this line seems to condtradict what you're saying: https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L78
> 
> But also, I'll include your comment in the doc directly, because it will be faster than me reading through gencred.

Well please hold off a few minutes if this is what you plan to do, I'll need to read through gencred code to make sure that I didn't miss something especially the RBAC stuff

Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md:58]: > Well please hold off a few minutes if this is what you plan to do, I'll need to read through gencred code to make sure that I didn't miss something especially the RBAC stuff

So according to https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/pkg/serviceaccount/serviceaccount.go#L36 that this KSA is a "cluster-admin" super user. I'm not entirely convinced that it's a requirement, feels like a security hole to me. @cjwagner , do you know what resource Prow needs to manage other than `Pod` resource on build clusters?

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md:16]: > could be named something totally unrelated

Where is this definition created? I want to add a link to an example.

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md]: done

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md]: done

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md:69]: done

File: site/content/en/docs/build-clusters.md
Comment (chaodaiG) [reply site/content/en/docs/build-clusters.md]: > > could be named something totally unrelated
> 
> Where is this definition created? I want to add a link to an example.

See https://github.com/kubernetes/test-infra/blob/408f66a2a9199fbcaafc55aa35f72a49a06da26c/gencred/cmd/gencred/main.go#L117, this is the `alias`

**Comment by listx** [comment]
Spoke offline with Chao and agreed to punt some of the additional details to a future PR if necessary. Can I get LGTM please?

**Review by chaodaiG — APPROVED** [review]

**Comment by chaodaiG** [comment]
Thank you Linus!

File: site/content/en/docs/build-clusters.md
Comment (listx) [reply site/content/en/docs/build-clusters.md]: Oh interesting. I was expecting a declarative YAML config somewhere, but it is defined in gencred first. So I: guess it's accurate to say that context aliases start out from gencred (makes sense to me, because kubeconfig seems to be the source of truth WRT how prow is able to talk to other build clusters, and as I now know, the token(s) in the kubeconfig need to be refreshed by gencred).

**Pull request merged**
